enhancements/substitutes-for.md

```diff
@@ -0,0 +1,379 @@
+
+1.0.0->1.0.1
+
+If the patched version is 1.0.1-patched (due to semver semantics this is ordered before 1.0.1), then ideally want want
+the new graph to be:
+
+[1.0.0 or 1.0.1-patched]->1.0.1
+
+However achieving such a graph would require updating the 1.0.1 bundle to add a "replaces 1.0.1-patched" reference.
+In turn, 1.0.1 can't be updated without changing its version, which means we end up needing something like:
+
+[1.0.0 or 1.0.1-patched or 1.0.1]->1.0.2
+
+(Where the only difference between 1.0.1 and 1.0.2 is the replaces references).
+
+
+To avoid having to update and publish new versions of subsequent bundles in the upgrade graph, we need a way to
+effectively stitch 1.0.1-patched into the graph without changing 1.0.1 (or any other existing bundles that need
+to be able to upgrade from 1.0.1-patched).
+
+This enhancement proposes we do so by allowing a bundle to specify which other bundles it "substitutes for".  This
+will effectively act as a search/replace in the existing bundles for any references to the bundles this bundle 
+substitutes for.  
+
+Continuing the example above, if 1.0.1-patch is annotated as substituting for 1.0.0, then the effective new upgrade 
+graph would be:
+
+1.0.0->1.0.0-patched->1.0.1
+
+
+The implications of this for opm add + the catalog db are that the substitutes-for metadata must be stored in the DB associated w/ each bundle.
+
+The expected flow is:
+
+1) opm add inserts the new bundle following all existing logic, but also adding the substitutes-for metadata
+2) Upon a list-bundles request, the operator-registry should return a substitution-corrected list of bundles based on substitutes replaces/skips references.
+3) In addition, the registry export command will need to perform substitutions on the CSVs as it exports them, to ensure their replaces/skips references are correct.
+
+
+Complications:
+
+1) the CSV itself contains other version references(beyond skips/replaces), but it should be ok not to perform updates on them because those references are not used by anything making upgrade decisions
+
+2) skipRange logic could theoretically get invalidated (1.2 substitutes for 1.1, but some bundle's skipRange is "skipRange: <1.2").  Since this feature is primarily being added for freshmaker which will only bump the patch release, this should simply be documented as a known limitation/caveat around using substitutes-for (that it does not allow substitution consideration during skipRange upgrades)
+
+3) If C substitutes B and B substituted A, it should be assumed that C substitutes A.  The db metadata and substitution processing needs to reflect this.
+
+
+## Motivation
+
+The primary motivation of this feature is to provide a way for the freshmaker tooling to rebuild a bundle, patching
+CVEs, and then add the new bundle to the operator catalog in such a way that clusters can upgrade to the
+new bundle from the unpatched bundle, as well as upgrade from the patched bundle to newer bundles.
+
+### Goals
```

Comment:
Should there be a goal for when a bundle author publishes a new bundle that replaces the unpatched version that opm identifies that this should actually be replacing the patched version?

Reply:
that should happen automatically with how this is designed.  either:

1) substitution happens during bundle add, in which case when the new bundle is added, all relevant substitutions are applied to it  (note this means whether the bundle being added is adding substitutions, or refs, or both, all of them must be evaluated against bundles that already exist in the catalog, potentially updating those existing bundles too)

or

2) substitution happens during listbundles, in which case again it'll be applied since the substitution is defined in the catalog

But yes, it's a key design point that you can at any time and in any order add additional references, or additional substitutions, and either way you get the same result, one that accounts for all the references and all the substitutions defined in the catalog.

Reply:
i think i misunderstood the intent of your question.  i'm adding this bullet explicitly:

* Future bundles that reference the unpatched version of a bundle, when a patched version exists in the
catalog, have their references updated to include the patched versions.


hopefully that covers it.